### DISCUSSION TITLE
Mosaic with Vega/Lite

### DISCUSSION THREAD

**Discussion by declann** [uwdata/mosaic#457] [General]
Hi,

> If you are interested in creating your own Mosaic clients, see the [Mosaic GitHub repository](https://github.com/uwdata/mosaic). For concrete examples, start with the source code of [Mosaic inputs](https://github.com/uwdata/mosaic/tree/main/packages/inputs/src). Once you've instantiated a custom client, register it using [coordinator.connect(client)](https://idl.uw.edu/mosaic/api/core/coordinator.html#connect).

I'm thinking about going down this route to make Vega visuals - perhaps eventually using signals for interactors going back to Mosaic.

Has anyone done this or come across examples like this?

The Vega runtime is fast 'enough' in my experience for interactive rendering responding to brushes etc. - when updating using the view API (e.g. [here](https://calculang.dev/examples-viewer?id=pension-calculator) responding to slider changes and [here](https://calculang-editables.netlify.app/tixyish) doing animation). And it has a tonne of flexibility for visuals and interactivity.

If this is new, could there be appetite to integrate something along these lines in a new example?

Thanks,
Declan

**Comment by domoritz** [comment]
There is a Vega example client in https://github.com/uwdata/mosaic/tree/main/packages/vega-example. Vega is well-equipped to build Mosaic clients since it can reactively update with very low overhead. We are also interested in building a generic Vega client where you just pass in the spec and it automatically extracts the transforms (similar to Vega fusion). The corresponding issue is at https://github.com/uwdata/mosaic/issues/397.

**Comment by declann** [comment]
That's an excellent example to work with and #397 is really interesting. Thanks Dominik

**Comment by jonmmease** [comment]
> We are also interested in building a generic Vega client where you just pass in the spec and it automatically extracts the transforms (similar to Vega fusion).

@domoritz Do you picture this working on a Vega spec, or a Vega-Lite spec? Seems like it would be pretty hard at the Vega level since the selection representations are pretty messy at that point.

**Reply by domoritz** [reply]
Vega-Lite makes more sense to start with but it could work for some subset of Vega as well.

**Reply by jheer** [reply]
I agree that Vega-Lite is a more attractive starting point. What remains to be seen is if the existing selection abstractions at the VL can stay the same with some extensions (eg for cross-filter resolution) or if they need to be modified.

**Reply by jonmmease** [reply]
@jheer Has anyone started working on reasoning through this (beyond the demo linked above)?
